### PR TITLE
Submodule support

### PR DESCRIPTION
**Pull request by nemui**
Hello, this is for #52

I'm not too sure how to correctly support submodules, to be honest, but this seems to be working for my use case: pushing tags on each build for the parent repository and its submodules, in other words, we rely on tags being the same across the repository.

What do you think?

I have also added a simple check for when toRef and fromRef are the same, #62  - in this case we return an empty list of tags.

### CODE REVIEW

**Comment by tomasbjerre** [comment]
I'll put this on my todo-list.

**Comment by nemui** [comment]
I have a better idea on how to handle submodules.
Instead of relying on specified refs to exist in all submodules, how about simply returning the diff between the current and the previous submodule hashes?

E.g. in the parent repo, we have 
commits A, B, C
And in a submodule we have commits D, E and F
A -> D
B -> D
C -> F

When diffing between A and C, we will also show the diff between D and F for commit C

**Comment by tomasbjerre** [comment]
Sounds good!

**Comment by nemui** [comment]
Great, I'll update the pull request once I have it working.

**Comment by nemui** [comment]
How does this look? I've opted out of having submodule changes per commit, as it makes the changelog too noisy, especially when mulitiple commits reference the same changes in submodules, e.g. via branch merges.

Instead, I diff submodules between the earliest and the latest commit mentioned in the parent repository for any given diffset.

**Comment by tomasbjerre** [comment]
This is on my todo-list. Will have a look later.

File: src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java [hunk 1]
```diff
@@ -380,6 +379,12 @@ public GitChangelogApi withSettings(final URL url) {
     return this;
   }
 
+  /** {@link Settings}. */
+  public GitChangelogApi withSettings(final Settings settings) {
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java:383]: I dont like to expose Settings in the API.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 2]
```diff
@@ -47,10 +41,12 @@
   private Git git;
   private final Repository repository;
   private final RevWalk revWalk;
+  private final HashMap<String, GitRepo> submodules;
 
   public GitRepo() {
     this.repository = null;
     this.revWalk = null;
+    this.submodules = null;
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:49]: I would prefer an empty map here, not null

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 3]
```diff
@@ -71,6 +67,26 @@ public GitRepo(final File repo) throws GitChangelogRepositoryException {
...
+              String submodulePath = submoduleWalk.getModulesPath();
+              this.submodules.put(submodulePath, new GitRepo(submoduleRepository.getDirectory()));
+            } catch (ConfigInvalidException e) {
+              LOG.warn("invalid submodule configuration; skipping submodule\n" + e);
+            }
+            submoduleRepository.close();
+          }
+        }
+      } else {
+        this.submodules = null;
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:87]: Should not be needed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 4]
```diff
@@ -89,6 +105,11 @@ public void close() throws IOException {
         LOG.error(e.getMessage(), e);
       }
     }
+    if (submodules != null) {
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:108]: This `if` is not needed, if changing to empty map as default value.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 5]
```diff
@@ -142,6 +163,45 @@ public ObjectId getRef(final String fromRef) throws GitChangelogRepositoryExcept
     throw new GitChangelogRepositoryException(fromRef + " not found in:\n" + toString());
   }
 
+  public boolean hasSubmodules() {
+    return submodules != null && submodules.size() > 0;
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:167]: This null-check is not needed, if changing to empty map as default value.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 6]
```diff
@@ -142,6 +163,45 @@ public ObjectId getRef(final String fromRef) throws GitChangelogRepositoryExcept
...
+
+  public String getDiff(String commitHash) {
+    RevCommit commit = null;
+    RevCommit prevCommit = null;
+
+    try {
+      commit = this.revWalk.parseCommit(getCommit(commitHash));
+      prevCommit = commit.getParentCount() > 0 ? commit.getParent(0) : null;
+    } catch (GitChangelogRepositoryException | IOException e) {
+      e.printStackTrace();
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:182]: throw GitChangelogRepositoryException

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 7]
```diff
@@ -142,6 +163,45 @@ public ObjectId getRef(final String fromRef) throws GitChangelogRepositoryExcept
...
+    DiffFormatter diffFormatter = new DiffFormatter(outputStream);
+    diffFormatter.setRepository(this.repository);
+    diffFormatter.setAbbreviationLength(10);
+
+    try {
+      for (DiffEntry entry : diffFormatter.scan(prevCommit, commit)) {
+        diffFormatter.format(diffFormatter.toFileHeader(entry));
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:195]: throw GitChangelogRepositoryException

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 8]
```diff
@@ -151,6 +211,21 @@ public String toString() {
...
+    RevTree revTree = commit.getTree();
+    if (revTree == null) {
+      return null;
+    }
+    ObjectId treeId = commit.getTree().getId();
+    ObjectReader reader = this.repository.newObjectReader();
+    try {
+      return new CanonicalTreeParser(null, reader, treeId);
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:224]: throw GitChangelogRepositoryException

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitSubmoduleParser.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+          submoduleEntries.put(submoduleName, submoduleEntry);
+        }
+
+        SubmoduleEntry existingEntry = submoduleEntries.getOrDefault(submoduleName, submoduleEntry);
+        existingEntry.previousSubmoduleHash = submoduleEntry.previousSubmoduleHash;
+      }
+    }
+
+    for (Map.Entry<String, SubmoduleEntry> submoduleEntry : submoduleEntries.entrySet()) {
+      settings.setFromCommit(submoduleEntry.getValue().previousSubmoduleHash);
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitSubmoduleParser.java:68]: I dont like changing the settings object, and later change it back. I would rather like a new instance created and used.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java [hunk 10]
```diff
@@ -291,6 +366,10 @@ private String getTagName(final Map<String, Ref> tagPerCommitHash, final String
     final RevCommit from = this.revWalk.lookupCommit(fromObjectId);
     final RevCommit to = this.revWalk.lookupCommit(toObjectId);
 
+    if (from == to) {
+      return newArrayList();
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:370]: If there is a tag on `from` then this list should not be empty.

**Comment by tomasbjerre** [comment]
This might work, but I have very limited time right now so need to be more complete. Needs test cases. The stuff that is added to `GitRepo` can probably be refactored out of there with a describing name for it.

File: src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java (hunk 1 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java:383]: Is there a better way to provide an existing Settings object?

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 2 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 3 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 5 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 6 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 7 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 8 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java]: Fixed.

**Comment by nemui** [comment]
Hi @tomasbjerre , I have added a couple of test cases.
> The stuff that is added to GitRepo can probably be refactored out of there with a describing name for it.
Something like GitRepoWithSubmodules class that inherits from GitRepo? Or is there an easier way to get access to repository and revWalk class members from GitRepo, as submodule functionality makes use of those?

File: src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java (hunk 10 above)
Comment (nemui) [reply src/main/java/se/bjurr/gitchangelog/internal/git/GitRepo.java:366]: Generally speaking - yes, but it looks like we are only interested in the diff between the two commits in this case?
If the tag of a single `from` commit, which is also the `to` commit, is returned, we get a non-empty diff, just like in the #62 description.

**Pull request closed without merging**